### ISSUE TITLE
[CII] VAT breakdown rounding problem, ex BR-Z-08 

### ISSUE THREAD

**Issue by tavi-cacina** [phax/validation#320]
For some inputs, the rule will wrongly be fired. Ex:
VAT BasisAmount: 1046.40
round sum LineTotal : 47212.32 (1)
round sum Allowance: 46165.92 (2)
(1) - (2) = 1046.4000000000014

The rule should round the whole right side of the equation. So, this:

`<param name="BR-Z-08" value="../ram:BasisAmount = (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'Z']/ram:SpecifiedTradeSettlementLineMonetarySummation/xs:decimal(ram:LineTotalAmount))*10*10)div 100) + (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[(ram:ChargeIndicator/udt:Indicator cast as xs:boolean)=true() and ram:CategoryTradeTax/ram:CategoryCode='Z']/xs:decimal(ram:ActualAmount))*10*10)div 100) - (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[(ram:ChargeIndicator/udt:Indicator cast as xs:boolean)=false() and ram:CategoryTradeTax/ram:CategoryCode='Z']/xs:decimal(ram:ActualAmount))*10*10)div 100)"/> `

should become:

`<param name="BR-Z-08" value="../ram:BasisAmount = round(((round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'Z']/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount)*10*10)div 100) + (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='true' and ram:CategoryTradeTax/ram:CategoryCode='Z']/ram:ActualAmount)*10*10)div 100) - (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='false' and ram:CategoryTradeTax/ram:CategoryCode='Z']/ram:ActualAmount)*10*10)div 100)) *10*10)div 100"/>  `

This may apply to other rules too, it is just that a client encountered it for this BR-Z-08

**Comment by phax** [comment]
Do you eventually have an example instance, in which this rules "misbehaves", so that we can analyze it?

**Comment by tavi-cacina** [comment]
Yes, here you are: [data.xml.txt](https://github.com/ConnectingEurope/eInvoicing-EN16931/files/9268334/data.xml.txt)

**Comment by MatthiasHainz** [comment]
There is a similiar rounding problem for this rule (BR-AE-08)

`<xsl:when test="../ram:BasisAmount = (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'AE']/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount)*10*10)div 100) + (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=true() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount)*10*10)div 100) - (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=false() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount)*10*10)div 100)" />`

Which should be

`<xsl:when test="../ram:BasisAmount = (round(xs:decimal(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'AE']/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount))*10*10)div 100) + (round(xs:decimal(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=true() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount))*10*10)div 100) - (round(xs:decimal(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=false() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount))*10*10)div 100)" />`

to avoid a false error

An example for a correct file which results in an error is here: 
[data.xml.txt](https://github.com/ConnectingEurope/eInvoicing-EN16931/files/9497614/data.xml.txt)

**Comment by phax** [comment]
Please note, in the above example, `*10*10` is displayed as *10*10
Please always try to use single back apostrophes to display code.
Inline code via "`" then the code and end with the same char
Code blocks starting with "```" and ending with the same

**Comment by phax** [comment]
Code blocks starting with "```xml" even get syntax highlighting but need to end without the "xml"

**Comment by phax** [comment]
```xml
<xsl:when test="../ram:BasisAmount = (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'AE']/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount)*10*10)div 100) + (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=true() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount)*10*10)div 100) - (round(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=false() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount)*10*10)div 100)" />
```

Which should be

```xml
<xsl:when test="../ram:BasisAmount = (round(xs:decimal(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'AE']/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount))*10*10)div 100) + (round(xs:decimal(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=true() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount))*10*10)div 100) - (round(xs:decimal(sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator=false() and ram:CategoryTradeTax/ram:CategoryCode='AE']/ram:ActualAmount))*10*10)div 100)" />
```

**Comment by oriol** [comment]
> Yes, here you are: [data.xml.txt](https://github.com/ConnectingEurope/eInvoicing-EN16931/files/9268334/data.xml.txt)

This example does not fail when tested with the current rule. Do you have other examples?

**Comment by phax** [comment]
@tavi-cacina ping
And a quick question: what XSLT processor are you using?

**Comment by tavi-cacina** [comment]
> This example does not fail when tested with the current rule.

Yes, you are right, it validates right with the current schema. Was my mistake, I tested accidentally with an old distribution, from itplr-kosit/validator-configuration-xrechnung: validator-configuration-xrechnung_2.0.1_2020-12-31. Current distributions have this rule fixed, is the `xs:decimal(ram:ActualAmount)` instead of plain `ram:ActualAmount` that avoids the problem.

> And a quick question: what XSLT processor are you using?

here was the second problem, we can not use the XSLT2 directly, it is not realy supported outside java world. We can not take the java runtime as a dependency (is also too slow as we validate interactively after each data-field change). So we changed the xslt so we can use it with an XSLT1 engine. This caused also the different behavior with the rounding, I thought is also a problem in upstream.
